### PR TITLE
flatpak: Don't propagate $TZDIR into sandbox

### LINKED ISSUE #238386: Kodi on flatpak show the wrong localtime
The time in kodi home page is always shown in UTC.
Kodi seem to use TZ env variable to configure the desired timezone (https://github.com/xbmc/xbmc/blob/8f774d10a07b14de5dd9f9984c4e77058bc4e185/xbmc/platform/posix/PosixTimezone.cpp#L190)
I tried in a bash shell inside container
> flatpak enter $(pidof kodi.bin) bash
>date
=> Sun Jun 18 11:31:13 CEST 2023
OK => CEST is 2h more than UTC
> TZ=CET date
=> Sun Jun 18 09:31:42 CET 2023
KO => should be Sun Jun 18 **11**:31:42 CE**S**T 2023
> TZ=CEST date
=> Sun Jun 18 09:31:45 CEST 2023
KO => should be Sun Jun 18 **11**:31:45 CE**S**T 2023

I tried those commands in another flatpak container (firefox), and the result is the same.
I tried it in kodi flatpak with another OS (ubuntu 22.04), and all the commands are working correctly (date in desired timezone)
I tried it directly on nixos, and the date is correctly shown with TZ=CET, but not with TZ=CEST.
I tried it directly on ubuntu 22.04, and all the commands are working correctly 

My nixos channel is https://nixos.org/channels/nixos-23.05

### PR DESCRIPTION
**Pull request by claes**
## Description of changes

This fixes https://github.com/NixOS/nixpkgs/issues/238386 by not propagating the TZDIR environment variable into the Flatpak sandbox. Since the value of TZDIR on NixOS is non-standard, and the directory does not exist within the sandbox, programs inside the sandbox that will attempt to access timezone files in TZDIR will fail and then produce wrong output.  

By exclusing TZDIR programs in the Flatpak sandbox will default to the standard default directory which is /usr/share/zoneinfo and where Flatpak org.freedesktop.Platform provides it.

Tested with Kodi

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#linking-nixos-module-tests-to-a-package) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) (or backporting [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) and [24.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

File: pkgs/by-name/fl/flatpak/package.nix
Comment (jtojnar) [inline pkgs/by-name/fl/flatpak/package.nix:125]: These patches should be kept, IMO, to bind mount the tzdata from host. Just blacklisting the `TZDIR` should be sufficient. Ideally, that would be done upstream.

Comment (claes) [reply pkgs/by-name/fl/flatpak/package.nix:125]: I think I understand now what Flatpaks attempt to do. But I don't think it works.

For example, when I set TZDIR to some custom directory where I add a test file, it is not present in the sandbox when running current nixpkgs 


```
sh-5.2$ mkdir /tmp/zoneinfo #NixOS host
sh-5.2$ cp -ar /etc/zoneinfo/* /tmp/zoneinfo
sh-5.2$ touch /tmp/zoneinfo/test #Create customer zoneinfo dir with test file
sh-5.2$ ./result/bin/flatpak  --env=TZDIR=/tmp/zoneinfo run tv.kodi.Kodi & 
[1] 63207
sh-5.2$ libva info: VA-API version 1.19.0
libva info: Trying to open /usr/lib/x86_64-linux-gnu/dri/iHD_drv_video.so
libva info: Trying to open /usr/lib/x86_64-linux-gnu/dri/intel-vaapi-driver/iHD_drv_video.so
libva info: Found init function __vaDriverInit_1_19
libva info: va_openDriver() returns 0
./result 
sh-5.2$ ./result/bin/flatpak enter $(pidof /app/lib/kodi/kodi.bin) bash
bash-5.2$ ls -l /usr/share/zoneinfo/test
ls: cannot access '/usr/share/zoneinfo/test': No such file or directory
```

Comment (jtojnar) [reply pkgs/by-name/fl/flatpak/package.nix:125]: Do you have flatpak from Nixpkgs with https://github.com/NixOS/nixpkgs/commit/801d1d4450f84070259c92f6bc292fe8e878bda3? It was only just merged.

Also not sure if `flatpak run --env` is the right flag – I would expect it to pass the environment variable to the guest, whereas Flatpak bind mounds tzdata based on `TZDIR` from the host.

Comment (claes) [reply pkgs/by-name/fl/flatpak/package.nix:125]: I have used getchoo's branch so it has been the right branch. But you were right when I set TZDIR like

`TZDIR=/tmp/zoneinfo  ./result/bin/flatpak   run tv.kodi.Kodi &` 

I can see my own contents under /usr/share/zoneinfo. So the Flatpak fix has an effect. I will put it back in

File: pkgs/by-name/fl/flatpak/unset-env-vars.patch
Comment (jtojnar) [inline pkgs/by-name/fl/flatpak/unset-env-vars.patch:9]: We may want to pick up https://github.com/flatpak/flatpak/pull/5942 once merged.

Comment (claes) [reply pkgs/by-name/fl/flatpak/unset-env-vars.patch:9]: It is my understanding it will go into the 1.15.X releases until 1.16 is released, which is when both this and the patches can be removed. Until then I think this works fine. It fixes the problem

**Review by getchoo — APPROVED** [review]

**Comment by claes** [comment]
@getchoo is there anything remaining for this to be merged?

**Comment by getchoo** [comment]
Not to my knowledge. We just need a committer :p

**Comment by claes** [comment]
Thank you @h7x4 !

**Pull request merged**
